ADV Add DaskWorker

rth opened the pull request:
This is a WIP attempt to implement a remote worker with dask.distributed.

Closes https://github.com/paris-saclay-cds/ramp-board/issues/444

PR to the `advanced` branch

## Usage

To use this worker, one would have to,
 1. setup a ramp event on the local machine, and deploy the event
 2. copy `ramp-kits` and `ramp-data` folder to the remote machine, install ramp-board and the appropriate conda env there. The structure of the `ramp_deployment` folder the remote machine should be the same as on the local machjne.
 3. Start a dask scheduler on the local machine, and one or several workers on the remote machine (cf https://docs.dask.org/en/latest/setup/cli.html#command-line) or use [dask-ssh](https://docs.dask.org/en/latest/setup/ssh.html) tool to do both.
 4. In the event config, specify
     ```json
     worker:
         worker_type: remote
         conda_env: ramp-iris
         dask_scheduler: "tcp://127.0.0.1:8786"   # or another URL to the dask scheduler
     ````
  5. Start the dispatcher for the event.

Some of the last 3 steps might be aggregated/automated.

## Implementation

The general idea is that we do exactly the same steps as in the local conda folder, but
 - we wrap calls into a `dask.distributed.Client.submit` to execute them on the remote machine
 - for data copies, getting logs etc we need to explicitly serialize it between remote/local using dask.

The nice thing is that we can run a local dask cluster, and easily run unit tests  on it. Currently I'm running all the  CondaEnvWorker unit tests and also dispatcher tests, both with this local worker and remote worker (there are still some failures in the dispatcher that should be fixed). All data serialization is not yet done via dask either.

There are some code changes in the `ramp_engine/local.py` as well, since I factorized functions that could be re-used in the remote worker in `ramp_engine/conda.py`

## TODO
 - [x] fix failing dispatcher tests with RemoteWorker
 - [x] adapt all file manipulation calls to use dask
 - [x] ~improve path adaption between the local and remote paths. E.g. in the event config we need to decide whether remote path or local paths are provided for each of the entries.~ to keep it simple, data, submission and kit paths should be the same between remote and local machine.
 - [x] test with an actual dask worker on a remote machine -> generally it work, a few tests still need fixing. 
 - [x] add documentation section

pep8speaks commented:
Hello @rth! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`ramp-engine/ramp_engine/remote.py`](https://github.com/paris-saclay-cds/ramp-board/blob/61e887aa91f65e5f6232cfc640ffe7b92dbcbbbd/ramp-engine/ramp_engine/remote.py):

> [Line 12:80](https://github.com/paris-saclay-cds/ramp-board/blob/61e887aa91f65e5f6232cfc640ffe7b92dbcbbbd/ramp-engine/ramp_engine/remote.py#L12): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 17:80](https://github.com/paris-saclay-cds/ramp-board/blob/61e887aa91f65e5f6232cfc640ffe7b92dbcbbbd/ramp-engine/ramp_engine/remote.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 83:80](https://github.com/paris-saclay-cds/ramp-board/blob/61e887aa91f65e5f6232cfc640ffe7b92dbcbbbd/ramp-engine/ramp_engine/remote.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

glemaitre commented:
We will be interested in master as well ;)

rth commented:
Yes, I know, there is nothing specific to the advanced branch here. I can look into backporting to master once it's merged here )

glemaitre commented:
> copy ramp-kits and ramp-data folder to the remote machine, install ramp-board and the appropriate conda env there. The structure of the ramp_deployment folder the remote machine should be the same as on the local machjne.

In the AWS worker, it is part of the worker to upload on the remote machine the submission and then collect the results.
Here, it seems that you want to outsource this part to the user. Would it make sense to have the same pattern?

glemaitre commented:
Another thing would be that it could be the job of the dispatcher to start a sask scheduler if requested, isn't it?

glemaitre commented on ramp-engine/ramp_engine/conda.py:
```diff
@@ -0,0 +1,68 @@
+from typing import Dict
+
+
+def _conda_info_envs() -> Dict:
```
I am not using `virtualenv` but do you think that it would be possible to have a higher-level API that could unify either using conda or virtualenv (of course not in this PR and anyway all these stuff are private).

rth replied:
yeah, there is nothing too conda specific in workers. We just need to compute the path to the python executable inside the virtualenv. The only issue is that virtualenv doesn't have a CLI to list all available envs. So we would probably need to use some additional tool on top, maybe [poetry](https://python-poetry.org/docs/managing-environments/#listing-the-environments-associated-with-the-project) or [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/command_ref.html?highlight=lsvirtualenv#lsvirtualenv) (unless we are OK passing the path to the virtualenv in the config).

rth commented:
> In the AWS worker, it is part of the worker to upload on the remote machine the submission and then collect the results.

The collection of the result is done in https://github.com/paris-saclay-cds/ramp-board/pull/452/files#diff-4f94babe188ee4f21f182143a33262de7bcd8ab89ba9520e9b67e1ca3f3f4a5bR261 but you are right that I still to add equivalent code for uploading submissions in https://github.com/paris-saclay-cds/ramp-board/pull/452/files#diff-4f94babe188ee4f21f182143a33262de7bcd8ab89ba9520e9b67e1ca3f3f4a5bR200 (there is a TODO note).

> Another thing would be that it could be the job of the dispatcher to start a dask scheduler if requested, isn't it?

Well it could (cf https://docs.dask.org/en/latest/setup/python-advanced.html#scheduler). Though that code is blocking, and would need to be run in an additional subprocess, and I'm not sure what would happen to stdout log handling. Then one need to start workers on the remote manually in any case, unless one uses [dask SSH](https://docs.dask.org/en/latest/setup/ssh.html). That's something we can investigate once we are sure that this fully works as expected, but initially starting them both manually is probably safer.

rth commented:
> In the AWS worker, it is part of the worker to upload on the remote machine the submission and then collect the results.

In the future, we could even probably copy the ramp-data and ramp-kit folder from the local machine, it's not present on the remote. But we probably don't want to do it for each submission (only once) for performance reasons, and I guess copying manually initially shouldn't be too problematic.

glemaitre commented:
> we could even probably copy the ramp-data and ramp-kit folder from the local machine

This is true that in AWS, we get this done by deploying the image.

OK so we will have to make a choice here at some point:

* get a common base class with high-level API shared between AWS and the current remote worker;
* or having dissociated class which will be more effective in terms of performance.

Anyway, this is not that important for the moment. Having something working is better and then we can see internal refactoring after seeing the limitation in practice.

glemaitre commented:
> Well it could (cf https://docs.dask.org/en/latest/setup/python-advanced.html#scheduler). Though that code is blocking, and would need to be run in an additional subprocess, and I'm not sure what would happen to stdout log handling. Then one need to start workers on the remote manually in any case, unless one uses dask SSH. That's something we can investigate once we are sure that this fully works as expected, but initially starting them both manually is probably safer.

I had a nice asynchronous non-blocking implementation for the dispatcher: https://gist.github.com/glemaitre/b0f0dddcadd9ebe7df39b657c777ac50 which was waiting on the shelves :)

> Then one need to start workers on the remote manually in any case, unless one uses dask SSH

OK so there is a little benefit there.

rth commented:
> I had a nice asynchronous non-blocking implementation for the dispatcher

Nice! Wouldn't it require for [workers to also be async](https://gist.github.com/glemaitre/b0f0dddcadd9ebe7df39b657c777ac50#file-asyncio-py-L44) though?

kegl commented:
The way we used AWS was to set up an image that contained everything except for the submission. Setting it up was semi-manual. We can have something formalized into some customizable script, what to do to set up the backend.

albertcthomas commented on ramp-engine/ramp_engine/remote.py:
```diff
@@ -0,0 +1,302 @@
+    def collect_results(self):
+        """Collect the results after that the submission is completed.
+
+        Be aware that calling ``collect_results()`` before that the submission
```
we could maybe add a check for this? if the worker.status is `running` we don't collect the results. If yes this can be refactored later maybe

rth replied:
Yes, that's what happens in the dispatcher, which handles when to call these methods depending on the worker status. However here this comment is still useful just to indicate the the call to `collect_results` is blocking.

albertcthomas commented on ramp-engine/ramp_engine/tests/test_conda_worker.py:
```diff
@@ -103,80 +115,96 @@ def test_conda_worker(submission, get_conda_worker):
 
+@pytest.mark.parametrize("Worker", ALL_WORKERS)
+def test_conda_worker_error_unknown_env(Worker, dask_scheduler):
+    with get_conda_worker(
```
maybe add a conda_env = 'xxx' variable used in the context manager declaration and in the msg_err

albertcthomas commented on ramp-engine/ramp_engine/tests/test_remote_worker.py:
```diff
@@ -0,0 +1,55 @@
+        'tcp://127.0.0.1:46663': 4
+    }
+
+    assert _check_dask_workers_single_machine(workers_nthreads.keys()) is True
```
I don't think you need the `is True`

rth replied:
without it, any not None or not False object would pass the assert. Here I wanted to check that the output is explicitly `True`

albertcthomas replied:
right

albertcthomas reviewed (commented):
I did not look at the dispatcher tests.

rth commented on ramp-engine/ramp_engine/remote.py:
```diff
@@ -0,0 +1,302 @@
+        shutil.rmtree(path, ignore_errors=True)
+
+
+class RemoteWorker(BaseWorker):
```
Good point. Renamed to `DaskWorker`, as it's not necessarily remote: it could be used with a local Dask cluster (as it's done in the tests).

rth commented:
Thanks a lot for the review @albertcthomas I addressed your comments.

> The way we used AWS was to set up an image that contained everything except for the submission. Setting it up was semi-manual. We can have something formalized into some customizable script, what to do to set up the backend.

Definitely, I agree @kegl we can certainly make this more user friendly. Both by using a pre-build Docker image for the worker and by automating more things (e.g. deployment of ramp-data, ramp-kit on the remote server. Or at least making sure they are consistent with the local server). As a first approach this works though, and I think, once I add a bit more docs, we can merge it into `advanced` since that's the version that's currently deployed for the event (and I would rather make PRs to advanced rather than this branch for other fixes).

agramfort commented on .gitignore:
```diff
@@ -25,4 +25,9 @@ test_submissions/submission*
+dask-worker-space/
+=======
 coverage.xml
+>>>>>>> upstream/master
```
Nice ;)

rth replied:
Thanks! Not sure how it got there, didn't see any conflicts..
